specification/hugr.md

```diff
@@ -1168,44 +1168,78 @@ The new hugr is then derived as follows:
 
 ###### `Replace`
 
-This takes as input:
+This is the general subgraph-replacement method.
 
-  - a set S of IDs of nodes in Γ, all of which are separated;
+A _partial hugr_ is is a graph $G$ having the same node and edge types as a hugr
+and satisfying all the same local constraints except that it may be missing edges.
 
-  - a partial hugr G;
+In particular:
 
-  - a map T from top(G) to IDs of container nodes in Γ\\S\*;
+  - it has no root node (the set of nodes without a parent is
+    denoted $\mathrm{top}(G)$ );
+  - it may have empty container nodes (the set of these is
+    denoted $\mathrm{bot}(G)$ ).
 
-  - a map B from bot(G) to IDs of container nodes in S\*, such that B(x)
-    is separated from B(y) unless x == y. Let X be the set of children
-    of values in B, and R be S\*\\X\*.
+A partial hugr describes a set of nodes and well-formed edges between
+them that potentially forms a subgraph of a hugr.
 
-  - a bijection μ<sub>inp</sub> between inp(G) and the set of input
-    ports of nodes in R whose source is not in R;
+Given a set $S$ of nodes in a hugr, let $S^\*$ be the set of all nodes
+descended from nodes in $S$ (i.e. reachable from $S$ by following hierarchy edges),
+including $S$ itself.
 
-  - a bijection μ<sub>out</sub> between out(G) and the set of output
-    ports of nodes in R whose target is not in R;
+Call two nodes $a, b \in \Gamma$ _separated_ if $a \notin \\{b\\}^\*$ and
+$b \notin \\{a\\}^\*$ (i.e. there is no hierarchy relation between them).
 
-  - disjoint subsets Init and Term of top(G);
+We define two structures to specify edge insertion.
 
-The new hugr is then derived by:
+A `NewInpEdgeSpec` specifies an edge inserted from an existing node to a (given) new node.
+It contains the following fields:
 
-1.  adding the new nodes from G;
+  - `EdgeKind`: may be `Value`, `Order`, `Static` or `ControlFlow`.
+  - `SrcNode`: the source node of the new edge.
+  - `SrcPort`: (for `Value` and `Static` edges only) the source port.
```

Comment:
Should we include this for control-flow also (as *which successor* matters wrt. unpacking of the "sum" type determining which way to branch)?

Or, is it possible that we do not need this operation to apply to CFGs? (A quick consideration suggests that we probably do, but I haven't thought hard about it)

Or perhaps this InpEdgeSpec cannot be a ControlFlow after all? Hmmm, it seems like we *do* need to specify what happens to each incoming CF edge, as simply removing a CF edge from the hugr will (generally speaking) invalidate it

Reply:
Ok, I go and read the spec ;). "Ports" are used only for Value/Static, whereupon we need both `SrcPort` and `DestPort`.

For ControlFlow, we need an integer position at the source, but not the target. For order, we don't need any position at either end.

So
* the source exists for Value, Static and Control Flow; it is a port for Value/Static, but an integer position for Controlflow. That might well be an int in all 3 cases.
* the dest exists for Value and Static only - as you write. But given we have "the source port" why not have "the dest port" similarly rather than "the desired position among the incoming ports").